handler/src/main/java/io/netty/handler/ssl/EnhancingX509ExtendedTrustManager.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2023 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.handler.ssl;
+
+import io.netty.util.internal.SuppressJava6Requirement;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.X509ExtendedTrustManager;
+import javax.net.ssl.X509TrustManager;
+import java.net.Socket;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.List;
+
+
+/**
+ * Wraps an existing {@link X509ExtendedTrustManager} and enhances the {@link CertificateException} that is thrown
+ * because of hostname validation.
+ */
+@SuppressJava6Requirement(reason = "Usage guarded by java version check")
+final class EnhancingX509ExtendedTrustManager extends X509ExtendedTrustManager {
+    private final X509ExtendedTrustManager wrapped;
+
+    EnhancingX509ExtendedTrustManager(X509TrustManager wrapped) {
+        this.wrapped = (X509ExtendedTrustManager) wrapped;
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket)
+            throws CertificateException {
+        try {
+            wrapped.checkClientTrusted(chain, authType, socket);
+        } catch (CertificateException e) {
+            throwEnhancedCertificateException(chain, e);
+        }
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket)
```

Comment:
We actually need to do this for `checkServerTrusted` instead of `checkClientTrusted`